Implement and use MultiArrayBuffer

geoffkizer opened the pull request:
MultiArrayBuffer is a struct that manages a sliding buffer, where bytes be added at the end and removed at the beginning, similar to existing ArrayBuffer. Unlike ArrayBuffer, it manages a list of fixed-size (16K) blocks internally.

This is most useful for scenarios where the buffer may grow large, and thus managing it as a single array is probably not efficient.

MultiArrayBuffer is now used in the HTTP2 response buffer code, as well as in StreamBuffer (used by ConnectedStreams and mock QUIC provider).

(Edit: This got moved to a separate PR, now merged) Also, change stream conformance tests to not use Assert.Equals for buffer comparsion, as it is very, very slow. This reduces the time of outerloop StreamConformanceTests by approximately 99%.

Edit: Also add AssertExtensions.SequenceEqual to provide more useful diff information when sequence comparison fails, and use it in stream conformance tests.

@stephentoub @dotnet/ncl

Commit:
add MultiArrayBuffer and use in StreamBuffer and Http2Stream

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+            // [initialBufferSize] is ignored for now;
+            // I kept it because some callers are passing useful info here that we might want to act on in the future.
+
+            _blocks = null;
+            _blockCount = 0;
+            _activeStart = 0;
+            _availableStart = 0;
```
Calling the default-ctor or `this = default` produces better zeroing code.
Don't know if it's worth it.

geoffkizer replied:
Maybe. I could add a factory method here that does the check, and change callers to use it. Let me look at it...

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+        {
+            get
+            {
+                if (index < 0 || index >= _length)
```
```suggestion
                if ((uint)index >= (uint)_length)
```
Saves one `cmp`.

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= ActiveMemory.Length, $"MultiArrayBuffer.Discard: Expected byteCount={byteCount} <= {ActiveMemory.Length}");
+
+            int oldStartBlock = _activeStart / BlockSize;
```
JIT produces optimized division for this. Casting to `uint`, then the JIT emits bit-shifts.
```suggestion
            int oldStartBlock = (int)((uint)_activeStart / BlockSize);
```
But code looks uglier.

The same on other places where `/ BlockSize`.

geoffkizer replied:
Yeah, I forgot that it will only optimize to a bit-shift if it's uint. I should probably change the fields here to be uint.

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+            {
+                Debug.Assert(start >= 0);
+                Debug.Assert(length >= 0);
+                Debug.Assert(start + length <= blocks.Length * BlockSize);
```
This could overflow.
`start + length` can overflow, as well as `blocks.Length * BlockSize`.

As it's `Debug.Assert` a `checked` doesn't harm.

gfoidl replied:
With the use of `uint` this is outdated.

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        private static int GetBlockIndex(int offset) => offset / BlockSize;
+        private static int GetOffsetInBlock(int offset) => offset % BlockSize;
```
Cast to `uint` produces the bit-hack modulo by using `and` instead of the optimized division.
```suggestion
        private static int GetOffsetInBlock(int offset) => (int)((uint)offset % BlockSize);
```

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+
+        public Memory<byte> GetBlock(int blockIndex)
+        {
+            if (blockIndex < 0 || blockIndex >= BlockCount)
```
```suggestion
            if ((uint)blockIndex >= BlockCount)
```

gfoidl replied:
Please take this one too.

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+
+        public MultiMemory Slice(int start)
+        {
+            if (start < 0 || start > _length)
```
```suggestion
            if ((uint)start > (uint)_length)
```

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+
+        public MultiMemory Slice(int start, int length)
+        {
+            if (start < 0 || length < 0 || start + length > _length)
```
`start + length` can overflow.

```suggestion
            if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
```

Or tho eke out more perf, one could (on 64-bit) bring it into `ulong`-domain and compare over there. But I don't think it's worth it.

<details>
  <summary>code</summary>

```c#
if (Environment.Is64BitProcess)
{
    if ((ulong)(uint)start + (ulong)(uint)length > (ulong)(uint)_length)
    {
        throw new IndexOutOfRangeException();
    }
}
else
{
    if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
    {
        throw new IndexOutOfRangeException();
    }
}
```
</details>

gfoidl reviewed (commented):
For the conditions with potential for overflow tests should be added? Or at least -- as it's internal -- asserts for them.

antonfirsov commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+
+        private const int BlockSize = 16 * 1024;
+
+        public MultiArrayBuffer(int initialinitialBufferSize)
```
```suggestion
        public MultiArrayBuffer(int initialBufferSize)
```

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+        // Ensure at least [byteCount] bytes to write to, up to the specified limit
+        public void TryEnsureAvailableSpaceUpToLimit(int byteCount, int limit)
+        {
+            if (ActiveMemory.Length >= limit)
```
Might be worth having an ActiveMemoryLength property that just computes the needed length, rather than creating an ActiveMemory with a length only to then unwrap it to get back the length.  Same for AvailableMemory.

geoffkizer replied:
Possibly. I've tried to err on the conservative side here, not adding stuff like that unless we determine we actually need it. The MultiMemory constructor is pretty cheap, so it seems like the JIT could inline it and optimize stuff away... that said, I don't know if it actually does this.

Looking at the places this is actually used, I think a more valuable change would probably be to avoid retrieving the ActiveMemory/AvailableMemory struct multiple times consecutively. Perhaps the JIT can optimize this too, but that doesn't seem super likely.

Thoughts?

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+        }
+
+        // Ensure at least [byteCount] bytes to write to, up to the specified limit
+        public void TryEnsureAvailableSpaceUpToLimit(int byteCount, int limit)
```
Every test of this method has limit == int.MaxValue.  What happens if you call it and the required byteCount would actually put you over the limit?  What does the Try prefix mean with no bool returned to indicate whether it was successful?

geoffkizer replied:
I think I will refactor this a bit, into two methods... one with a limit, one without. If you don't care about a limit then we can avoid a small amount of work here.

And I'll add tests specifically for these methods.

"Try" is a legacy here and I'll drop it.

> What happens if you call it and the required byteCount would actually put you over the limit?

Then we will allocate up to the limit and no more. So if you call this, it's your responsibility to check how much space was made available, since it may be less than you asked for, if you hit the limit.

You could argue that the caller could enforce this themselves, I suppose. This mattered more on ArrayBuffer, since we wanted to limit the size of the actual byte array if you have a limit. But here we are basically just using it to do a Math.Min.

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+                    blockArraySize *= 2;
+                }
+
+                _blocks = new byte[]?[blockArraySize];
```
Is it worth renting this from a pool, too?

stephentoub replied:
@jaredpar, I still struggle with our having reversed for nullability the positions of the array specifiers here :(  It's so confusing.

geoffkizer replied:
Possibly. I'm not sure how well the pool does with small allocations, any thoughts?

The good news is that there will hopefully be significant reuse here if you have lots of MultiArrayBuffers, since we start with 4 entries here and then double from there. 

The bad news is that it seems pretty unlikely that anyone else is pooling arrays of byte[]...

stephentoub replied:
> Possibly. I'm not sure how well the pool does with small allocations, any thoughts?

Currently the smallest bucket size is 16, so any request for smaller will return a length 16 array.

stephentoub replied:
> The bad news is that it seems pretty unlikely that anyone else is pooling arrays of byte[]

No other code in the core libraries is.

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+            // Allocate new blocks
+            for (int i = 0; i < newBlocksNeeded; i++)
+            {
+                _blocks[_blockCount + i] = ArrayPool<byte>.Shared.Rent(BlockSize);
```
This seems a bit aggressive, but I guess it's logically no different than doubling buffer space available, it just happens to be split across N buffers?

geoffkizer replied:
How is it aggressive? We are allocating as many new blocks as needed to satisfy the user's request for space, no more.

stephentoub replied:
Sorry, you're right; I saw a `* 2` earlier and in my head thought that was for number of blocks.

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+
+        public int Length => _length;
+
+        public ref byte this[int index]
```
With perf measurements, it might be worth aggressively inlining this, at which point we'd also want to move the exception throw into a throw helper method.

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,319 @@
+
+        public int BlockCount => _length == 0 ? 0 : GetBlockIndex(_start + _length - 1) - GetBlockIndex(_start) + 1;
+
+        public Memory<byte> GetBlock(int blockIndex)
```
Any reason this needs to return a `Memory<byte>` rather than `Span<byte>`?

geoffkizer replied:
No strong reason, just that Memory is more usable than Span. E.g. I could pass it into an async method if I wanted to.

stephentoub commented on src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:
```diff
@@ -905,7 +905,7 @@ public virtual async Task Read_PopulatedWithInitialData_ToEof_Success(ReadWriteM
             }
 
-            Assert.Equal(expected, actual.ToArray());
+            Assert.True(expected.AsSpan().SequenceEqual(actual.ToArray()));
```
```suggestion
            AssertExtensions.Equal(expected, actual.ToArray());
```

geoffkizer replied:
I'm happy to make this change, since it seems like `AssertExtensions.Equal` is the correct method to use.

That said, I really wish this method was named differently, e.g. `AssertExtensions.SequenceEquals` or something. As it stands, it's not obvious from looking at the code that it's actually doing the right thing here (unlike `Assert.Equals`). And I never would have found this on my own without you pointing me to it -- I actually did look on AssertExtensions, but was expecting something like SequenceEquals or SpanEquals or whatever, not just Equal.

stephentoub commented on src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:
```diff
@@ -990,7 +990,7 @@ public virtual async Task Write_CustomMemoryManager_Success(bool useAsync)
             byte[] actual = (byte[])expected.Clone();
             Assert.Equal(actual.Length, await ReadAllAsync(ReadWriteMode.AsyncMemory, stream, actual, 0, actual.Length));
-            Assert.Equal(expected, actual);
+            Assert.True(expected.AsSpan().SequenceEqual(actual));
```
```suggestion
            AssertExtensions.Equal(expected, actual);
```

stephentoub commented on src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:
```diff
@@ -1032,7 +1032,7 @@ public virtual async Task Write_CustomMemoryManager_Success(bool useAsync)
             }
 
-            Assert.Equal(expected.AsSpan(position).ToArray(), destination.ToArray());
+            Assert.True(expected.AsSpan(position).SequenceEqual(destination.ToArray()));
```
```suggestion
            AssertExtensions.Equal(expected.AsSpan(position).ToArray(), destination.ToArray());
```

stephentoub commented on src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:
```diff
@@ -1073,7 +1073,7 @@ public virtual async Task Write_Read_Success(ReadWriteMode mode)
             {
                 int bytesRead = await ReadAllAsync(mode, stream, actual, 0, actual.Length);
-                Assert.Equal(expected, actual);
+                Assert.True(expected.AsSpan().SequenceEqual(actual));
```
```suggestion
                AssertExtensions.Equal(expected, actual);
```

stephentoub commented on src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:
```diff
@@ -1686,7 +1686,7 @@ public virtual async Task ReadWriteByte_Success()
                 }
 
-                Assert.Equal(writerBytes, readerBytes);
+                Assert.True(writerBytes.AsSpan().SequenceEqual(readerBytes));
```
```suggestion
                AssertExtensions.Equal(writerBytes, readerBytes);
```

stephentoub commented on src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:
```diff
@@ -2367,7 +2367,7 @@ public virtual async Task CopyToAsync_AllDataCopied(int byteCount, bool useAsync
             await copyTask;
 
-            Assert.Equal(dataToCopy, results.ToArray());
+            Assert.True(dataToCopy.AsSpan().SequenceEqual(results.ToArray()));
```
```suggestion
            AssertExtensions.Equal(dataToCopy, results.ToArray());
```

geoffkizer commented:
Lots of stream conformance failures... I didn't see these locally. Will investigate.

I'm going to factor out the span compare change into a separate PR, so we can get that in and confirm that's not an issue here.

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

En3Tho commented:
@stephentoub Just a side question - I see more and more API's using ArrayPool.Shared. And my question is: aren't we going to get pool drained quickly this way (considering user code too)? Or are there any changes to Shared pool implementation? Afaik this pool is not endless if I may call it like that

stephentoub commented:
I use the term "drained", but it's actually a bit misleading, in that it suggests there's a fixed number of buffers that are available in the pool and if they're gone they're gone and the pool is no longer effective... that's not actually how it works.  Today each size bucket in the pool has a max number of buffers that it can store, but they're not pre-generated: if you go to the pool and there isn't a buffer available, it manufactures a new one for you.  It's on returning a buffer to the pool that it might get thrown away if the relevant bucket has already reached its threshold. And this is important because it means you have can way, way more buffers of a given size "rented" at any one time than the bucket's threshold suggests, and as long as they're not all returned to live in the pool at the same time, many of them won't be thrown away: the pool ends up just being a brief hand-off point between one component adding a buffer and another one coming along and taking one. Strange analogy, but kind of like a bathroom at a restaurant: you don't need one bathroom per customer, rather you can be feeding many, many more customers than you have bathrooms, and you only need as many bathrooms as you want to enable customers to use them at the same time.

Which is to say, yes, it's possible that as more and more components use the ArrayPool we'll need to update its thresholds and behaviors to accomodate the demand, but we've not yet seen it to be a problem. https://github.com/dotnet/runtime/issues/12800 tracks revisiting some of this.

En3Tho commented:
@stephentoub I see. Thanks for the detailed explanation. I've just been concerned that at some point excessive use of array pool might become an "expensive new" in unexpected ways.

Commit:
rework EnsureAvailableSpace and add tests for it

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

davidfowl commented:
This feels like something that should be public. Pipelines does a similar thing and uses a linked list instead of an array to manage segments. It's the writable side of a ReadOnlySequence.

stephentoub commented:
> I've just been concerned that at some point excessive use of array pool might become an "expensive new" in unexpected ways.

It's definitely possible. It's why I opened #12800 to keep on eye on it and revisit various aspects of it as necessary. 

> This feels like something that should be public.

That's the end goal.  For now it needs to be proven out, both implementation and surface area, via multiple consumers.  Having pipelines switch to using it would help with both aspects.

Commit:
make dispose handling more robust, plus minor tweaks

Commit:
add AssertExtensions.SequenceEqual for providing more useful information about sequence differences, and use in StreamConformanceTests

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

geoffkizer commented:
/azp help

geoffkizer commented:
/azp list

geoffkizer commented:
/azp run runtime

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

geoffkizer commented:
@dotnet/dnceng I keep getting cancelled jobs here. How do I investigate these? All I see in the logs is that the Send To Helix job takes a long time and is then cancelled.

lpatalas commented:
We'll take a look at it.

Commit:
tweaks

Commit:
replace _blockCount with _allocatedEnd

Commit:
use uint instead of int internally to assure bitop optimizations

Commit:
fix _allocatedEnd calculation

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,377 @@
+
+            // Allocate new blocks
+            Debug.Assert(_allocatedEnd % BlockSize == 0);
+            uint allocatedBlockCount = _allocatedEnd / BlockSize;
```
```suggestion
            uint allocatedBlockCount = _allocatedEnd / BlockSize;
            Debug.Assert(allocatedBlockCount == 0 || _blocks[allocatedBlockCount - 1] is not null);
```
?

geoffkizer replied:
Added

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,377 @@
+        {
+            if (destination.Length < _length)
+            {
+                throw new ArgumentException(nameof(destination));
```
I'm surprised this isn't triggering an analyzer warning.  This is using the argument name as the exception message (ArgumentException uses annoyingly inconsistent ctor arguments.)

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,377 @@
+        {
+            if (_length < source.Length)
+            {
+                throw new ArgumentException(nameof(source));
```
Ditto

stephentoub commented on src/libraries/Common/src/System/Net/StreamBuffer.cs:
```diff
@@ -40,7 +40,7 @@ public bool IsComplete
                 lock (SyncObject)
                 {
-                    return (_writeEnded && _buffer.ActiveLength == 0);
+                    return (_writeEnded && _buffer.ActiveMemory.Length == 0);
```
Maybe MultiArrayBuffer should have an IsEmpty property.

stephentoub commented on src/libraries/Common/src/System/Net/StreamBuffer.cs:
```diff
@@ -277,9 +277,9 @@ public void AbortRead()
 
                 _readAborted = true;
-                if (_buffer.ActiveLength != 0)
+                if (_buffer.ActiveMemory.Length != 0)
```
Nit:  A few lines above in the diff we have a `> 0` check, and here we have a `!= 0` check.  There's no difference, is there?

geoffkizer replied:
Nope, just inconsistency. Will fix.

stephentoub commented on src/libraries/Common/src/System/Net/StreamBuffer.cs:
```diff
@@ -292,7 +292,10 @@ public void Dispose()
             EndWrite();
 
-            _buffer.Dispose();
+            lock (SyncObject)
```
Was the lack of this previously a bug, or is there something about switching to use MultiArrayBuffer that requires this?

geoffkizer replied:
It was previously a bug, but the window was tiny. With MultiArrayBuffer, the window is a bit bigger.

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:
```diff
@@ -378,6 +380,53 @@ public static void Equal<T>(HashSet<T> expected, HashSet<T> actual)
+                }
+                else
+                {
+                    const int maxDiffsToShow = 10;      // arbitrary; enough to be useful, hopefully, but still manageable
```
Nit:
```suggestion
                    const int MaxDiffsToShow = 10;      // arbitrary; enough to be useful, hopefully, but still manageable
```

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:
```diff
@@ -378,6 +380,53 @@ public static void Equal<T>(HashSet<T> expected, HashSet<T> actual)
+                    const int maxDiffsToShow = 10;      // arbitrary; enough to be useful, hopefully, but still manageable
+
+                    int diffCount = 0;
+                    string message = $"Showing first {maxDiffsToShow} differences{Environment.NewLine}";
```
```suggestion
                    string message = $"Showing first {MaxDiffsToShow} differences{Environment.NewLine}";
```

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs:
```diff
@@ -378,6 +380,53 @@ public static void Equal<T>(HashSet<T> expected, HashSet<T> actual)
+                            diffCount++;
+
+                            // Add up to 10 differences to the exception message
+                            if (diffCount <= maxDiffsToShow)
```
```suggestion
                            if (diffCount <= MaxDiffsToShow)
```

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -99,7 +99,7 @@ public Http2Stream(HttpRequestMessage request, Http2Connection connection)
                 _responseProtocolState = ResponseProtocolState.ExpectingStatus;
 
-                _responseBuffer = new ArrayBuffer(InitialStreamBufferSize, usePool: true);
+                _responseBuffer = new MultiArrayBuffer(InitialStreamBufferSize);
```
I don't know if it's going to matter, but this does mean we're now allocating one or more byte[][]s per HTTP/2 stream, whereas previously all of the arrays here were from the pool.

geoffkizer replied:
Yes.

We could pool the block array too, if we wanted to; I'm just not sure whether that is worthwhile or not.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -1280,7 +1280,10 @@ private void CloseResponseBody()
                 }
 
-                _responseBuffer.Dispose();
+                lock (SyncObject)
```
Same lock/dispose question as earlier.

geoffkizer replied:
See above.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -1045,10 +1045,10 @@ private void ExtendWindow(int amount)
-                        int bytesRead = Math.Min(buffer.Length, _responseBuffer.ActiveLength);
-                        _responseBuffer.ActiveSpan.Slice(0, bytesRead).CopyTo(buffer);
+                        int bytesRead = Math.Min(buffer.Length, _responseBuffer.ActiveMemory.Length);
+                        _responseBuffer.ActiveMemory.Slice(0, bytesRead).CopyTo(buffer);
```
Store ActiveMemory into a local rather than accessing it three times?

geoffkizer commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,377 @@
+            {
+                for (int i = 0; i < _blocks.Length; i++)
+                {
+                    if (_blocks[i] is not null)
+                    {
+                        ArrayPool<byte>.Shared.Return(_blocks[i]!);
+                        _blocks[i] = null;
+                    }
```
Will do.

I assumed that ArrayPool.Return would not throw. Is that incorrect?

stephentoub replied:
It shouldn't,  I'm mostly being paranoid.  But for example if tracing is enabled it could end allocating and OOM'ing.

Commit:
address PR feedback

Commit:
more PR feedback and some minor fixes/improvements

geoffkizer commented:
I addressed all the above issues and made a couple minor improvements. In particular, I added DiscardAll and IsEmpty, and added CheckState for debug builds that validates the state of the buffer array.

PTAL

geoffkizer commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,377 @@
+                Debug.Assert(_blocks[oldStartBlock] is not null, $"Discard: oldStartBlock is null?? byteCount={byteCount}, _activeStart={_activeStart}, oldStartBlock={oldStartBlock}, newStartBlock={newStartBlock}");
+
+                ArrayPool<byte>.Shared.Return(_blocks[oldStartBlock]!);
+                _blocks[oldStartBlock] = null;
```
Done

geoffkizer commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,377 @@
+
+            while (oldStartBlock < newStartBlock)
+            {
+                Debug.Assert(_blocks is not null);
```
Done

geoffkizer commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,377 @@
+        public MultiArrayBuffer(int initialBufferSize)
+        {
+            // [initialBufferSize] is ignored for now;
+            // I kept it because some callers are passing useful info here that we might want to act on in the future.
```
Done

geoffkizer commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -136,83 +140,99 @@ public void GrowAvailableSpace(int byteCount)
 
                 int blockArraySize = 4;
-                while (blockArraySize < newBlocksNeeded)
+                while (blockArraySize  < newBlocksNeeded)
```
Fixed, thanks

Commit:
small test improvements

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

stephentoub reviewed (commented):
Skimmed the last few commits: LGTM.

Have you done any perf measurements on HTTP/2 to confirm this doesn't regress anything?

gfoidl commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+            else
+            {
+                Debug.Assert(blocks is not null);
+                Debug.Assert(start >= 0);
+                Debug.Assert(length >= 0);
```
With `uint` these will always be true.
Should check instead
```suggestion
                Debug.Assert(start <= int.MaxValue);
                Debug.Assert(length <= int.MaxValue);
```
?

(Should the compiler have shown a suggestion for this?)

geoffkizer commented:
> Have you done any perf measurements on HTTP/2 to confirm this doesn't regress anything?

No, I haven't. I expect it doesn't make much difference for small responses, whereas for large responses, we know we have a problem and moving to a multi-buffer approach is a necessary part of the solution.

geoffkizer committed:
Apply suggestions from code review

Co-authored-by: Günther Foidl <gue@korporal.at>

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop

geoffkizer commented:
@dotnet/ncl

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+    // Unlike ArrayBuffer, the buffer itself is managed using 16K blocks which are added/removed to the block list as necessary.
+
+    // [ActiveBuffer] contains the current buffer contents; these bytes will be preserved on any call to TryEnsureAvailableBytesUpToLimit.
+    // [AvailableBuffer] contains the available bytes past the end of the current content,
```
Nit: rather than using braces like `[ActiveBuffer]`, how about using single quotes or backticks, e.g. `'ActiveBuffer'`?  The braces just immediately make me think we're talking about attributes.

geoffkizer replied:
Sure

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+            _allocatedEnd = 0;
+            _activeStart = 0;
+            _availableStart = 0;
+        }
```
Nit: it'd be a bit more efficient to do:
```C#
public MultiArrayBuffer(int initialBufferSize) : this()
{
    // [initialBufferSize] is ignored for now. Some callers are passing useful info here that we might want to act on in the future.
    Debug.Assert(initialBufferSize >= 0);
}
```

geoffkizer replied:
Fixed

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+                    }
+                    else
+                    {
+                        // We can shift the array down to make enough space
```
Did you consider making the byte[][] circular rather than shifting like this?

geoffkizer replied:
Yes, that seems like a potentially good optimization. I elected not to do it for now.

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+        {
+            uint ustart = (uint)start;
+            uint ulength = (uint)length;
+            if (ustart > _length || ustart + ulength > _length)
```
Is the first check needed?

geoffkizer replied:
No, good catch

geoffkizer replied:
Ah, actually, I remember why I did this.

We aren't explicitly checking here if `start >= 0`, but if it is, this will get caught in the `ustart > _length` test.

Unfortunately the second check here isn't quite right, since it doesn't catch a case like start = 5, length = -1, I think it should instead be `ulength > _length - ustart` and that should correctly catch all invalid cases.

This is also why we don't have asserts here, as you asked about above.

Note also there are asserts in the MultiMemory constructor, so if we miss something here it will get asserted there.

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+        }
+
+        public MultiMemory Slice(int start)
+        {
```
`Debug.Assert(start >= 0);`?

geoffkizer replied:
See comments above

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+            return new MultiMemory(_blocks, _start + ustart, _length - ustart);
+        }
+
+        public MultiMemory Slice(int start, int length)
```
`Debug.Assert(start >= 0 && length >= 0);`?

geoffkizer replied:
See comments above

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+                throw new ArgumentOutOfRangeException(nameof(destination));
+            }
+
+            for (int blockIndex = 0; blockIndex < BlockCount; blockIndex++)
```
BlockCount looks like it does some computation.  Store it into a local rather than recomputing it for each iteration?

stephentoub commented on src/libraries/Common/src/System/Net/MultiArrayBuffer.cs:
```diff
@@ -0,0 +1,431 @@
+                throw new ArgumentOutOfRangeException(nameof(source));
+            }
+
+            for (int blockIndex = 0; blockIndex < BlockCount; blockIndex++)
```
Ditto

Commit:
address review feedback

geoffkizer commented:
/azp run runtime-libraries-coreclr outerloop
